### ISSUE TITLE
CMB2 as a dependency - loads too late

### ISSUE THREAD

**Issue by maxchirkov** [WebDevStudios/CMB2#181]
CMB2 is initiated by a version-appended class, which changes on every version, so there is no way to consistently check if init class exists (before it was cmb2_bootstrap_200beta and now it's cmb2_bootstrap_201). The earliest consistent class that loads is CMB2, which loads on init hook with very low priority (9999). This means it executes after most plugins already ran their own init methods. In cases where one would like to use CMB2 as a dependency, it becomes very inconvenient to check if CMB2 exists, since that plugin's action hooks would have to be shifted either into the lower priority of init or even after it. This limits flexibility of the init hook to plugins that depend on CMB2.

The solution would be to make it easier to check if CMB2 exists early in the loading process: 
1. either stop changing the bootstrap class; 
2. add global flag indicating that CMB exists, 
3. or load CMB2 class right away so it could be detected by class_exists() early, and simply delay it's instantiation to the init hook.

Thoughts?

**Comment by jtsternberg** [comment]
The only solution that will work while maintaining the mechanism we have in place to ensure only the newest version of CMB2 is loaded is your 2nd bullet. I think we could pretty easily add a `CMB2_LOADED` constant in the init class's constructor that you could check. Also, a good hook to use (in place of the `init` hook) if you want to hook after CMB2 is loaded, is the [`cmb2_init` hook](https://github.com/WebDevStudios/CMB2/blob/master/CMB2.php#L30).

**Comment by maxchirkov** [comment]
Thanks for the prompt response.

A global constant sounds good to me. As far as cmb2_init hook goes, it kinda makes things backwards - a library fires up an application bootstrap and not the other way around. In my case, I have different controllers that have their own init hooks, which will never get executed if plugin loads on cmb2_init.

Thanks again!

**Comment by jtsternberg** [comment]
To be clear, I agree, it IS a bit backwards, and it comes down to the issue that dependency management in WordPress is broken. For now, this is the best we got. I'd be open to further discussion, but keep in mind, CMB2 can be installed as a plugin, or bundled in a plugin/theme or placed in mu-plugins. It needs to be intelligent enough to handle all those scenarios without failing, and still load the most recent version installed in the system.

**Comment by mAAdhaTTah** [comment]
> CMB2 can be installed as a plugin, or bundled in a plugin/theme or placed in mu-plugins.

While maybe it's not possible to retrofit CMB2, I've always wondered whether it really makes sense for these options libraries to support all of these options. CMB2 in particular has no front-end display if you don't code it up yourself, so it never really made sense to me why an options library would support being installed as a plugin, other than convention. Was this an active decision or just conforming to standard practice?

**Comment by maxchirkov** [comment]
Don't get me wrong, I'm not criticizing :) I'm grateful you made this available to the rest of us.

**Comment by jtsternberg** [comment]
@mAAdhaTTah having it be installable as a plugin is the least of our concerns as far as this dependency management issue goes. But to answer your questions, we've made CMB2 available as a plugin for several reasons:

* Easy 'installation' (simply activate the plugin)
* Free support forum to funnel people to (instead of cluttering issues w/ support requests)
* More visibility and less barrier to entry
* Why NOT have it installable as a plugin? It works the same no matter where you put it.

@maxchirkov Nah, I completely agree with you. I love to hate CMB2. It's the nature of a software maintainer.

**Comment by mAAdhaTTah** [comment]
> Why NOT have it installable as a plugin? It works the same no matter where you put it.

Mostly related to this, as supporting all these methods can conflict. But if that has nothing to do with this issue, then nvm.

**Comment by jtsternberg** [comment]
@mAAdhaTTah the issue comes when CMB2 is loaded/included from more than one source (plugin or theme or mu-plugins, or wp plugin). The mechanisms in place are to ensure that CMB2 being included from more than one place will not break/white-screen. Again, dependency management is hard. :(

**Comment by maxchirkov** [comment]
Thank you very much!

**Comment by apintocr** [comment]
Just to leave my opinion :)

I believe plugins and libraries should be ALWAYS installed as a plugin (or mu-plugin).
This avoids duplication of data in case 2 plugins (or themes) want the same resource.

Also, I believe it is the correct workflow.

For better WordPress dependency management, try composer, it is amazing (wp-packagist :+1: )
